### PR TITLE
Add authorization and authentication with JWT

### PR DESCRIPTION
**Pull request by N0menIllisLegio**
Changed DB structure to store RefreshTokens, to update DB you need to run this script:
```
USE [BookSharing];

CREATE TABLE [RefreshTokens] (
	[ID] INT IDENTITY(1,1) NOT NULL PRIMARY KEY,
	[Token] VARCHAR(250) NOT NULL,
	[ExpirationDate] DATETIME NOT NULL,
	[AccountID] INT NOT NULL,

	CONSTRAINT [FK_RefreshTokens_AccountID] FOREIGN KEY([AccountID])
		REFERENCES [Accounts]([ID])
		ON UPDATE CASCADE
		ON DELETE CASCADE
);
```

I think we need to add presentation layer to project structure.

### CODE REVIEW

File: BookSharing/Controllers/AuthController.cs [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+
+namespace BookSharing.Controllers
+{
+    [Route("api/[controller]/[action]")]
+    public class AuthController : Controller
+    {
+        private readonly BookSharingContext _context;
+        private readonly IAuthUtility _authUtility;
+
+        public AuthController(BookSharingContext context, IAuthUtility authUtility)
```
Comment (s-samusik) [inline BookSharing/Controllers/AuthController.cs:22]: let's remove the underscore in front of the private variables, okay?
It seems to me that this is not useful now, but only complicates the reading of the code.

File: BookSharing/Controllers/AuthController.cs [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+using System.Linq;
+using System.Security.Claims;
+using System.Threading.Tasks;
+
+namespace BookSharing.Controllers
+{
+    [Route("api/[controller]/[action]")]
+    public class AuthController : Controller
+    {
+        private readonly BookSharingContext _context;
```
Comment (s-samusik) [inline BookSharing/Controllers/AuthController.cs:19]: `private readonly BookSharingContext _context;`

We don't have to implement a database connection here.
We can create instance of AccountRepository and get Account uses public method "GetAccountByQuery(string query)" 
where query - our email (Today i am implementing this method).

File: BookSharing/Controllers/AuthController.cs [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+                    Email = model.Email,
+                    Password = model.Password,
+
+                    FirstName = model.FirstName,
+                    LastName = model.LastName,
+                    PhoneNumber = model.PhoneNumber,
+                    AccountType = await _context.AccountTypes
+                        .FirstAsync(at => at.Name == "User")
+                };
+
```
Comment (s-samusik) [inline BookSharing/Controllers/AuthController.cs:75]: Here we can call methods:
public Task<Account> CreateAccountAsync(all parameters) - today i am implementing this method.
and 
public Task AddAccountAsync(Account account)

There will be no business logic here.
This controller should only receive requests, send to the logical layer, and return a response to the presentation layer.

**Review by s-samusik — COMMENTED** [review]
Tell me please why "RefreshToken.cs" located in BookSharing.Domain/Models/, but
"SignInModel.cs, SignUpModel.cs, TokenRequestModel.cs, TokenResponseModel.cs"
located in BookSharing.Domain/Data/Auth/Models/

This controller BookSharing/Controllers/AuthController.cs will become simple

"I think we need to add presentation layer to project structure."
What do you mean? The presentation layer is our MVC project. Or I misunderstood something?

**Comment by N0menIllisLegio** [comment]
> Tell me please why "RefreshToken.cs" located in BookSharing.Domain/Models/, but
> "SignInModel.cs, SignUpModel.cs, TokenRequestModel.cs, TokenResponseModel.cs"
> located in BookSharing.Domain/Data/Auth/Models/

They are not located in DAL because there is no such models in DB. They are representing requests from client.

> This controller BookSharing/Controllers/AuthController.cs will become simple. There will be no business logic here.
> This controller should only receive requests, send to the logical layer, and return a response to the presentation layer.

Ok.

> "I think we need to add presentation layer to project structure."
> What do you mean? The presentation layer is our MVC project. Or I misunderstood something?

You often say that all logic should be in BLL but in solution we have: *Domain* project which consists, if I am undestanding correctly, BLL and DAL, *Test* project that (should) have tests and *BookSharing* and there lays PL (again if I am guessing correctly). I think for readablilty there should be 3 project DAL, BLL, PL + Tests.

> And i think we need coverage xUnitTests for BookSharing.Domain/Data/Auth/AuthUtility.cs

Actually we need it everywhere but do we  have time for it?

File: BookSharing/Controllers/AuthController.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (N0menIllisLegio) [reply BookSharing/Controllers/AuthController.cs:22]: Ok

File: BookSharing/Controllers/AuthController.cs (hunk 2 above)
Comment (N0menIllisLegio) [reply BookSharing/Controllers/AuthController.cs:19]: Than I will wait

File: BookSharing/Controllers/AuthController.cs (hunk 3 above)
Comment (N0menIllisLegio) [reply BookSharing/Controllers/AuthController.cs:75]: I got it

**Comment by s-samusik** [comment]
> You often say that all logic should be in BLL but in solution we have: Domain project which consists, if I am undestanding correctly, BLL and DAL, Test project that (should) have tests and BookSharing and there lays PL (again if I am guessing correctly). I think for readablilty there should be 3 project DAL, BLL, PL + Tests.
Yes you are right.

The structure of the solution can be done in different ways, I spent three evenings studying this))
Some options are simpler, others more difficult.
Now I propose this option and I think that it suits us for our tasks (if you change it, then perhaps you need to do all over again)

About solution's structure:
project BookSharing - (Presentation layer)

project Booksharing.Domain (Business layer)
 - folder "Data" (realization of interfaces or concrete level)
 - folder "Interfaces" (abstract level)
 - folder "Models" (DAL level and i suggest DTO level)
 - folder "Services" (more complex logic of concrete level of business layer)

project Booksharing.Tests
 (testing only business layer concrete level of Domain project)
>Actually we need it everywhere but do we have time for it?

I think it’s a not good idea to leave a project without tests. We will write checks for basic conditions (absence, incorrect data).
It does not take much time. For example, I write tests, you improve the authorization module ))

File: BookSharing.Domain/Data/AuthRepository.cs [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        }
+
+        public async Task<ResponseModel> SignUpAsync(SignUpModel user)
+        {
+            var account = new Account
+            {
+                Nickname = user.Nickname,
+                Email = user.Email,
+                Password = authUtility.Sha256(user.Password),
+                // Password = user.Password, where we hashing pass? here?
```
Comment (s-samusik) [inline BookSharing.Domain/Data/AuthRepository.cs:29]: I think here or we may create Service (for example "PasswordService") and here call some method from this service.

File: BookSharing.Domain/Interfaces/IAccountRepository.cs [hunk 5]
```diff
@@ -58,5 +58,19 @@ public interface IAccountRepository
...
+        /// <returns>Object "AccountType" or null.</returns>
+        Task<AccountType> GetAccountTypeByNameAsync(string typeName);
+
+        Task AddNewRefreshTokenAsync(RefreshToken refreshToken);
+
+        Task<RefreshToken> GetRefreshTokenAsync(string email, string refreshToken);
+
+        Task DeleteRefreshTokenAsync(RefreshToken refreshToken);
+
+        Task DeleteAllRefreshTokensFromAccountAsync(Account account);
```
Comment (s-samusik) [inline BookSharing.Domain/Interfaces/IAccountRepository.cs:74]: Don't like descriptions for methods? ))
```
Task AddNewRefreshTokenAsync(RefreshToken refreshToken);
Task<RefreshToken> GetRefreshTokenAsync(string email, string refreshToken);
Task DeleteRefreshTokenAsync(RefreshToken refreshToken);
Task DeleteAllRefreshTokensFromAccountAsync(Account account);
```

File: BookSharing.Domain/Interfaces/IAuthRepository.cs [hunk 6]
```diff
@@ -0,0 +1,14 @@
+﻿using BookSharing.Domain.Data.Auth.Models;
+using System.Threading.Tasks;
+
+namespace BookSharing.Domain.Interfaces
+{
+    public interface IAuthRepository
+    {
```
Comment (s-samusik) [inline BookSharing.Domain/Interfaces/IAuthRepository.cs:7]: Don't like descriptions for methods? ))

**Review by s-samusik — APPROVED** [review]

**Pull request merged**